[JSC][32-bit] Optimize test32 and compare32 with branchless code generation

mikhailramalho opened the pull request:
#### 3baaaee049d934750e4ac547eb41a2d3eb421df1
<pre>
[JSC][32-bit] Optimize test32 and compare32 with branchless code generation
<a href="https://bugs.webkit.org/show_bug.cgi?id=302016">https://bugs.webkit.org/show_bug.cgi?id=302016</a>

Reviewed by Justin Michaud.

This PR replace IT blocks with branchless instruction sequences for common
test32 and compare32 operations.

This PR also removes NOPs in conditional moves via condition inversions, and
adds assertions for invalid Carry/Overflow conditions (tst doesn&apos;t set these
flags).

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::invert):
(JSC::MacroAssemblerARMv7::commuteCompareToZeroIntoTest):
(JSC::MacroAssemblerARMv7::branch32):
(JSC::MacroAssemblerARMv7::compare32):
(JSC::MacroAssemblerARMv7::test32):
(JSC::MacroAssemblerARMv7::moveConditionally32):
(JSC::MacroAssemblerARMv7::moveConditionallyTest32):
(JSC::MacroAssemblerARMv7::moveDoubleConditionally32):
(JSC::MacroAssemblerARMv7::moveDoubleConditionallyTest32):
(JSC::MacroAssemblerARMv7::moveDoubleOrNop): Deleted.
(JSC::MacroAssemblerARMv7::moveOrNop): Deleted.

Canonical link: <a href="https://commits.webkit.org/303329@main">https://commits.webkit.org/303329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6bea7c8f2104001e39af3a391e0df52985b9060

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139373 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83792 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4407 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100837 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68223 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134807 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3139 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118101 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3033 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/810 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82593 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123924 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142016 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130368 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4022 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36805 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109165 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4103 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3565 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109331 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27731 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3056 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114382 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57239 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4076 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32782 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163334 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3908 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67523 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42465 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4036 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
```diff
@@ -3310,42 +3468,114 @@ class MacroAssemblerARMv7 : public AbstractMacroAssembler<Assembler> {
     void moveConditionallyTest32(ResultCondition cond, RegisterID left, TrustedImm32 right, RegisterID thenCase, RegisterID elseCase, RegisterID dest)
     {
+        if (thenCase == elseCase) {
+            move(thenCase, dest);
```
Maybe leaving the comment here would be helpful, just to keep it in mind

mikhailramalho replied:
done

mikhailramalho committed:
[JSC][32-bit] Optimize test32 and compare32 with branchless code generation
https://bugs.webkit.org/show_bug.cgi?id=302016

Reviewed by Justin Michaud.

This PR replace IT blocks with branchless instruction sequences for common
test32 and compare32 operations.

This PR also removes NOPs in conditional moves via condition inversions, and
adds assertions for invalid Carry/Overflow conditions (tst doesn't set these
flags).

* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::invert):
(JSC::MacroAssemblerARMv7::commuteCompareToZeroIntoTest):
(JSC::MacroAssemblerARMv7::branch32):
(JSC::MacroAssemblerARMv7::compare32):
(JSC::MacroAssemblerARMv7::test32):
(JSC::MacroAssemblerARMv7::moveConditionally32):
(JSC::MacroAssemblerARMv7::moveConditionallyTest32):
(JSC::MacroAssemblerARMv7::moveDoubleConditionally32):
(JSC::MacroAssemblerARMv7::moveDoubleConditionallyTest32):
(JSC::MacroAssemblerARMv7::moveDoubleOrNop): Deleted.
(JSC::MacroAssemblerARMv7::moveOrNop): Deleted.

Canonical link: https://commits.webkit.org/303329@main
